HeterogeneousCore/AlpakaInterface/README.md

```diff
@@ -4,3 +4,310 @@ This package only depends on the `alpaka` header-only external library, and
+one-dimensional (array) alpaka host buffers:
+
+  - `host_buffer<T>` (where `T` is not an array)
+    represents a single object of type `T`;
+
+  - `host_buffer<T[N]>`
+    represents a one-dimensional buffer of `N` objects of type `T`;
+
+  - `host_buffer<T[]>`
+    represents a one-dimensional buffer of a number of objects of type `T`
+    determined at construction time.
+
+
+### `const` host memory buffers
+
+As for an `std::shared_ptr`, the use of a `const` buffer is not particularly
+usefeful, because one can always make a non-`const` copy that shares ownership
+of the same underlying memory and objects.
+When `const`-only access to a buffer is desired, an `alpaka::ViewConst` adapter
+can be used to provide the required semantics.
+
+These type aliases implement host buffers with `const`-only semantics:
+
+  - `const_host_buffer<T>`;
+  - `const_host_buffer<T[N]>`;
+  - `const_host_buffer<T[]>`.
+
+A `const_host_buffer` can be automatically constructued from a `host_buffer`.
+These are mostly useful as function arguments, to guarantee that the function
+does not make changes to the underlying memory.
+
+
+### Allocating host buffers
+
+These helper functions allocate zero-dimensional (scalars) and one-dimensional
+(array) alpaka buffers in general purpose host memory:
+
+  - `auto make_host_buffer<T>()` (where `T` is not an array)
+    allocates a zero-dimensional buffer holding a single object of type `T`;
+
+  - `auto make_host_buffer<T[N]>()`
+    allocates a one-dimensional buffer holding `N` objects of type `T`;
+
+  - `auto make_host_buffer<T[]>(size)`
+    allocates a one-dimensional buffer holding `size` objects of type `T`.
+
+These function allocate the buffers in general purpose host memory, using
+immediate operations, without any caching.
+
+The memory is not initialised.
+
+### Allocating host buffers in device-mapped memory
+
+These helper functions allocate zero-dimensional (scalars) and one-dimensional
+(array) alpaka buffers in device-mapped memory:
+
+  - `auto make_host_buffer<T, Platform>()` (where `T` is not an array)
+    allocates a zero-dimensional buffer holding a single object of type `T`;
+
+  - `auto make_host_buffer<T[N], Platform>()`
+    allocates a one-dimensional buffer holding `N` objects of type `T`;
+
+  - `auto make_host_buffer<T[], Platform>(size)`
+    allocates a one-dimensional buffer holding `size` objects of type `T`.
+
+These functions allocate the buffers in host memory that is pinned and mapped in
+the address space of the devices, using potentially blocking operations, without
+any caching. The memory and objects can be accessed both on the host and on the
+devices.
+
+The memory is not initialised.
+
+On most systems accessing this memory from the device is significantly slower
+than using device global memory, so it is not adviced to use these buffers
+directly in device kernels. A common use case for device-mapped memory is as a
+staging area for copies to and from the devices.
```

Comment:
Would it be worth of mentioning somewhere here that the following cases require device-host synchronization (since these pitfalls do not apply to the device buffers)?

* after `alpaka::memcpy(queue, ...)` from host to device (or device kernels that would access the host memory directly, even if that was discouraged above), the source host buffer can be written into again only after a synchronization
* after `alpaka::mempcy(queue, ...)` from device to host, the destination host buffer can be accessed (read or write) only after a synchronization

While these are sort of obvious, I feel mentioning them explicitly could be useful for newcomers

Reply:
Good points, thanks for mentioning them.

I am also thinking that a `host_buffer` with pinned memory _could_ be used for communication between the GPU and CPU (_e.g._ the GPU could write there a size, and the CPU could later use it to allocate a buffer sized accordingly); however this is quite advanced and would require some _ad-hoc_ host-device synchronisation, so maybe this would be outside the scope of this documentation ?

Reply:
> I am also thinking that a `host_buffer` with pinned memory _could_ be used for communication between the GPU and CPU (_e.g._ the GPU could write there a size, and the CPU could later use it to allocate a buffer sized accordingly); however this is quite advanced and would require some _ad-hoc_ host-device synchronisation, so maybe this would be outside the scope of this documentation ?

Yeah, I feel this pattern is something that should be tried out first before documenting it.